File: 17.电话号码的字母组合.c
/*
 * @lc app=leetcode.cn id=17 lang=c
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
char **allStr;
char *item;
char **result;
int top1;
int top2;
int top3;

void back(int k, int startIndex)
{
  if (!k)
    return;
  if (top2 == k)
  {
    int i;
    char *path = (char *)malloc(sizeof(char) * (k + 1));
    for (i = 0; i < k; i++)
      path[i] = item[i];
    path[i] = '\0';
    result[top3++] = path;
    return;
  }

  int i, j;
  for (i = startIndex; i < top1; i++)
  {
    for (j = 0; j < strlen(allStr[i]); j++)
    {
      item[top2++] = allStr[i][j];
      back(k, i + 1);
      top2--;
    }
  }
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
char **letterCombinations(char *digits, int *returnSize)
{
  int i, start, len = strlen(digits);
  result = (char **)malloc(sizeof(char *) * 1000);
  allStr = (char **)malloc(sizeof(char *) * len);
  item = (char *)malloc(sizeof(char) * (len + 1));
  top1 = top2 = top3 = 0;
  for (i = 0; i < len; i++)
  {
    if (digits[i] == '8' || digits[i] == '9')
      start = 'a' + 3 * (digits[i] - '0' - 2) + 1;
    else
      start = 'a' + 3 * (digits[i] - '0' - 2);
    if (digits[i] == '7' || digits[i] == '9')
    {
      char *temp = (char *)malloc(sizeof(char) * 5);
      temp[0] = start;
      temp[1] = start + 1;
      temp[2] = start + 2;
      temp[3] = start + 3;
      temp[4] = '\0';
      allStr[top1++] = temp;
    }
    else
    {
      char *temp = (char *)malloc(sizeof(char) * 5);
      temp[0] = start;
      temp[1] = start + 1;
      temp[2] = start + 2;
      temp[3] = '\0';
      allStr[top1++] = temp;
    }
  }
  back(len, 0);
  *returnSize = top3;
  return result;
}
// @lc code=end

File: store/17.电话号码的字母组合.js
/*
 * @lc app=leetcode.cn id=17 lang=javascript
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  if (!digits) return [];
  const arr = [
    0,
    0,
    ["a", "b", "c"],
    ["d", "e", "f"],
    ["g", "h", "i"],
    ["j", "k", "l"],
    ["m", "n", "o"],
    ["p", "q", "r", "s"],
    ["t", "u", "v"],
    ["w", "x", "y", "z"],
  ];
  let res = [],
    item = [];
  const backtracking = (arr, digits, start) => {
    if (item.length === digits.length) {
      res.push(item.join(""));
      return;
    }
    // 法一
    // for (let i = start; i < digits.length; i++) {
    //   for (let j = 0; j < arr[digits[i]].length; j++) {
    //     item.push(arr[digits[i]][j]);
    //     backtracking(arr, digits, i + 1);
    //     item.pop();
    //   }
    // }
    // 法二
    for (let j = 0; j < arr[digits[start]].length; j++) {
      item.push(arr[digits[start]][j]);
      backtracking(arr, digits, start + 1);
      item.pop();
    }
  };

  backtracking(arr, digits, 0);

  return res;
};
// @lc code=end
